src/compiler/objective_c_generator.cc

```diff
@@ -176,65 +176,71 @@ void PrintMethodImplementations(Printer *printer,
 
 string GetHeader(const ServiceDescriptor *service, const string prefix) {
   string output;
-  grpc::protobuf::io::StringOutputStream output_stream(&output);
-  Printer printer(&output_stream, '$');
+  {
+    // Scope the output stream so it closes and finalizes output to the string.
```

Comment:
>  we're taking addresses of temporaries

You mean the string? Taking the address of an automatic variable is ok, if you don't use that address beyond the variable's life.

If `Printer` expects the `OutputStream` to be alive during `~Printer`, then we actually can't have both being local variables in the same scope, as there's no reason the compiler should call the destructors in the order we expect. I now think so far we've just been lucky, and the additional scope is the way to satisfy Printer's contract. If OutputStream has the same requirement for the string pointer it receives, then we actually need two scopes IMO. Ew.

> There's no way moving the string before calling the local destructors is an acceptable behavior.

Well if the compiler wants to move it at all, it has to do it before calling the local destructors, or that will destruct the string too _before_ moving it. I would bet the reason we're not seeing it elsewhere is the compilers being conservative and not moving, plus string's copy constructor not copying the actual characters.